### PR TITLE
Adopt msal for non-managed identity credentials

### LINKED ISSUE #14920: Convert `azidentity` to use msal/go
Switch the innards to use msal/go instead of our hand-written implementation.

### LINKED ISSUE #15632: InteractiveBrowserCredentialOptions.RedirectURL doesn't work
`InteractiveBrowserCredential.GetToken()` starts a redirect server only when `InteractiveBrowserCredentialOptions.RedirectURL` has no value:
https://github.com/Azure/azure-sdk-for-go/blob/6ff991256abe3ac3c5b5c43a977f34f230782a02/sdk/azidentity/interactive_browser_credential.go#L128-L132

Acquiring a token is therefore impossible when the option has a value. Fixing this is more complex than simply ensuring the redirect server starts, because the server doesn't work for all valid values of `RedirectURL`, e.g. `https://127.0.0.1`. However, MSAL for Go supports `RedirectURL`, so this should be easy to fix as part of #14920.

### LINKED ISSUE #15973: Add live test recordings
For the current live tests, and also add new tests for managed identity.

### LINKED ISSUE #16118: Remove azcore.ClientOptions copying
After #16091, it isn't necessary for credential constructors to copy `azcore.ClientOptions` parameters, because `runtime.NewPipeline` protects them from mutation.

### PR DESCRIPTION
**Pull request by chlowell**
Removes aad_identity_client.go and reimplements the credentials that were using it with MSAL clients. Adds recorded tests for these credentials so we don't lose all our test coverage. Closes #14920, closes #15632, closes #15973, closes #16118

### CODE REVIEW

File: sdk/azidentity/go.mod [hunk 1]
```diff
@@ -2,12 +2,13 @@ module github.com/Azure/azure-sdk-for-go/sdk/azidentity
 
 go 1.16
 
+replace github.com/Azure/azure-sdk-for-go/sdk/azcore => ../azcore
```
Comment (jhendrixMSFT) [inline sdk/azidentity/go.mod:5]: Do we need this?

File: sdk/azidentity/go.mod [hunk 2]
```diff
@@ -2,12 +2,13 @@ module github.com/Azure/azure-sdk-for-go/sdk/azidentity
...
+
 require (
 	github.com/Azure/azure-sdk-for-go/sdk/azcore v0.20.0
 	github.com/Azure/azure-sdk-for-go/sdk/internal v0.8.2
+	github.com/AzureAD/microsoft-authentication-library-for-go v0.3.1
```
Comment (jhendrixMSFT) [inline sdk/azidentity/go.mod:10]: v0.4.0 is the latest preview that supports OBO

File: sdk/azidentity/go.mod (hunk 1 above)
Comment (chlowell) [reply sdk/azidentity/go.mod:5]: Yes, until we release a version with `runtime.PipelineOptions`

Comment (jhendrixMSFT) [reply sdk/azidentity/go.mod:5]: Ah I see.  So I assume we won't be releasing this azidentity change until that time.

Comment (chlowell) [reply sdk/azidentity/go.mod:5]: That's my plan. I thought, why write code that's already out of date vs. main when there's no urgency to release azidentity.

File: sdk/azidentity/interactive_browser_credential_test.go [hunk 3]
```diff
@@ -5,15 +5,16 @@ package azidentity
...
 	"testing"
+	"time"
 
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
-	"github.com/Azure/azure-sdk-for-go/sdk/internal/mock"
-	"golang.org/x/net/http2"
+	"github.com/AzureAD/microsoft-authentication-library-for-go/apps/public"
 )
 
+var runManualBrowserTests = os.Getenv("AZIDENTITY_RUN_MANUAL_BROWSER_TESTS") != ""
```
Comment (seankane-msft) [inline sdk/azidentity/interactive_browser_credential_test.go:16]: If you're checking to make sure the value is set you could use 
`var _, runManualBrowserTests = os.LookupEnv("...")`

**Review by seankane-msft — APPROVED** [review]

File: sdk/azidentity/azidentity.go [hunk 4]
```diff
@@ -28,30 +42,13 @@ const (
...
-	headerUserAgent   = "User-Agent"
-	headerURLEncoded  = "application/x-www-form-urlencoded"
-	headerMetadata    = "Metadata"
-	headerContentType = "Content-Type"
-)
-
-const tenantIDValidationErr = "Invalid tenantID provided. You can locate your tenantID by following the instructions listed here: https://docs.microsoft.com/partner-center/find-ids-and-domain-names."
-
 var (
 	successStatusCodes = [2]int{
```
Comment (jhendrixMSFT) [inline sdk/azidentity/azidentity.go:46]: Is this still needed?

File: sdk/azidentity/azidentity.go [hunk 5]
```diff
@@ -80,15 +77,60 @@ func validTenantID(tenantID string) bool {
...
-// oauthPath returns the oauth path for AAD or ADFS
-func oauthPath(tenantID string) string {
-	if tenantID == "adfs" {
-		return "/oauth2"
+func (p pipelineAdapter) CloseIdleConnections() {
+	// do nothing
+}
+
+func (p pipelineAdapter) Do(r *http.Request) (*http.Response, error) {
+	req, err := runtime.NewRequest(r.Context(), r.Method, r.URL.String())
```
Comment (jhendrixMSFT) [inline sdk/azidentity/azidentity.go:94]: It's unfortunate that we have to do this (I feel like I had to do this elsewhere too).  Maybe we need a `runtime.WrapRequest` that creates a `runtime.Request` from an existing `http.Request`.

File: sdk/azidentity/authorization_code_credential.go [hunk 6]
```diff
@@ -44,32 +47,70 @@ func NewAuthorizationCodeCredential(tenantID string, clientID string, authCode s
...
+	return &AuthorizationCodeCredential{authCode: authCode, clientSecret: options.ClientSecret, redirectURI: redirectURL, pca: c}, nil
 }
 
 // GetToken obtains a token from Azure Active Directory by redeeming the authorization code. This method is called automatically by Azure SDK clients.
 // ctx: Context controlling the request lifetime.
 // opts: Options for the token request, in particular the desired scope of the access token.
 func (c *AuthorizationCodeCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (*azcore.AccessToken, error) {
-	tk, err := c.client.authenticateAuthCode(ctx, c.tenantID, c.clientID, c.authCode, c.clientSecret, "", c.redirectURI, opts.Scopes)
+	if c.cca != nil {
+		ar, err := c.cca.AcquireTokenSilent(ctx, opts.Scopes, confidential.WithSilentAccount(c.account))
```
Comment (jhendrixMSFT) [inline sdk/azidentity/authorization_code_credential.go:86]: I bet you could place this common code in a local func.
```go
acquireToken := func(func acquireSilent(/*params*/), func acquireByAuthCode(/*params*/)) (msal.AuthenticationResult, error) {
// stuff....
}
var ar msal.AuthenticationResult
var err error
if c.cca != nil {
    ar, err = acquireToken(c.cca.AcquireTokenSilent, c.cca.AcquireTokenByAuthCode)
}
// etc...
```

**Review by jhendrixMSFT — APPROVED** [review]

File: sdk/azidentity/authorization_code_credential.go (hunk 6 above)
Comment (chlowell) [reply sdk/azidentity/authorization_code_credential.go:86]: The sticky part is that the MSAL methods have incompatible options params, e.g. `public.AcquireTokenSilent` requires `public.AcquireTokenSilentOption`, so a function which might call `confidential.AcquireTokenSilent` instead must know the difference. I guess I could hide the difference behind another type but imagine that isn't worth the cost in legibility. What do you think of something like this:
```go
var acquireTokenSilent func() (base.AuthResult, error)
if c.cca != nil {
    acquireTokenSilent = func() (base.AuthResult, error) {
        return c.cca.AcquireTokenSilent(ctx, opts.Scopes, confidential.WithSilentAccount(c.account))
    }
} else {
    acquireTokenSilent = func() (base.AuthResult, error) {
        return c.pca.AcquireTokenSilent(ctx, opts.Scopes, public.WithSilentAccount(c.account))
    }
}
```
Still repetitive but less so, and allows sharing the business logic at least.

Comment (jhendrixMSFT) [reply sdk/azidentity/authorization_code_credential.go:86]: Ah good point, I didn't look super close.  What you have is fine.

**Pull request merged**